zcash_client_backend/src/tor/http/cryptex.rs

```diff
@@ -0,0 +1,174 @@
+    ///
+    /// Gemini is treated as a "trusted" data source due to being a NYDFS-regulated
+    /// exchange.
+    pub fn unauthenticated_known_with_gemini_trusted() -> Self {
+        Self::builder(exchanges::Gemini::unauthenticated())
+            .with(exchanges::Binance::unauthenticated())
+            .with(exchanges::Coinbase::unauthenticated())
+            .with(exchanges::GateIo::unauthenticated())
+            .with(exchanges::KuCoin::unauthenticated())
+            .with(exchanges::Mexc::unauthenticated())
+            .build()
+    }
+
+    /// Returns an `Exchanges` builder.
+    ///
+    /// The `trusted` exchange will always have its data used, _if_ data is successfully
+    /// obtained via Tor (i.e. no transient failures).
+    pub fn builder(trusted: impl Exchange) -> ExchangesBuilder {
+        ExchangesBuilder::new(trusted)
+    }
+}
+
+/// Builder type for [`Exchanges`].
+pub struct ExchangesBuilder(Exchanges);
+
+impl ExchangesBuilder {
+    /// Constructs a new [`Exchanges`] builder.
+    ///
+    /// The `trusted` exchange will always have its data used, _if_ data is successfully
+    /// obtained via Tor (i.e. no transient failures).
+    pub fn new(trusted: impl Exchange) -> Self {
+        Self(Exchanges {
+            trusted: Box::new(trusted),
+            others: vec![],
+        })
+    }
+
+    /// Adds another [`Exchange`] as a data source.
+    pub fn with(mut self, other: impl Exchange) -> Self {
+        self.0.others.push(Box::new(other));
+        self
+    }
+
+    /// Builds the [`Exchanges`].
+    pub fn build(self) -> Exchanges {
+        self.0
+    }
+}
+
+impl Client {
+    /// Fetches the latest USD/ZEC exchange rate, derived from the given exchanges.
+    ///
+    /// Returns:
+    /// - `Ok(rate)` if at least one exchange request succeeds.
+    /// - `Err(_)` if none of the exchange queries succeed.
+    pub async fn get_latest_zec_to_usd_rate(
+        &self,
+        exchanges: &Exchanges,
+    ) -> Result<Decimal, Error> {
+        // Fetch the data in parallel.
+        let res = join!(
+            exchanges.trusted.query_zec_to_usd(self),
+            join_all(exchanges.others.iter().map(|e| e.query_zec_to_usd(self)))
+        );
+        trace!(?res, "Data results");
+        let (trusted_res, other_res) = res;
+
+        // Split into successful queries and errors.
+        fn split(s: &mut Vec<Decimal>, e: &mut Vec<Error>, res: Result<ExchangeData, Error>) {
+            match res {
+                Ok(d) => s.push(d.exchange_rate()),
+                Err(error) => e.push(error),
+            }
+        }
+        let mut rates: Vec<Decimal> = vec![];
+        let mut errors = vec![];
+        for res in other_res {
+            split(&mut rates, &mut errors, res);
+        }
+
+        // "Never go to sea with two chronometers; take one or three."
+        // Randomly drop one rate if necessary to have an odd number of rates.
```

Comment:
Should an error be returned if exactly two successful responses are returned? If only one successful response is returned, should it be used if it's not the trusted exchange? Should we even permit using only a single source of truth?

Reply:
Now that we allow the caller to configure the exchanges, we _have_ to permit a single source of truth, because the builder allows them to construct this (it requires 1 but does not require others, and the one required source may be set to a source of data that they themselves control). However, we might want to prohibit configuring only two sources of truth: if any other source is provided, require providing at least two other sources. That means making the builder fallible in a way that would force almost every normal usage to `unwrap`, which feels wrong.

If exactly two successful responses are returned, _and we have more than two sources_, then I would be instead inclined to retry the others if there's a chance it was a transient failure. Determining what is a transient failure is hard however, because "the exit node used for this specific query is blocked / rate limited" is in our case a transient failure, whereas "all Tor exit nodes are blocked" is not.

Reply:
Okay, I agree that since we only require one source to be configured, we should allow single-source responses. I think that we shouldn't contort the API to prohibit configuring only two sources of truth; however, maybe if three or more sources are configured, then we should require at least three responses, and return an error otherwise?

Reply:
> maybe if three or more sources are configured, then we should require at least three responses, and return an error otherwise?

At least 3 responses, or should 1 also be acceptable and we only error on 2?

For this PR I'm not going to stick in retries, but we can add that later if we determine that "error on 2" is causing too many problems.

> It's still not clear to me what the behavior should be if two sources are configured; probably take the trusted source if available, and treat the second source as a backup?

That seems reasonable and comprehensible.

Reply:
If 3 are configured, require 3 is best I think. If you're content with a single source, just configure 2. It's a small hole in the API, but I think it's sensible.